core/txpool/bundlepool/bundlepool.go

```diff
@@ -0,0 +1,442 @@
+	bundleSlotSize = 128 * 1024 // 128KB
+)
+
+var (
+	bundleGauge = metrics.NewRegisteredGauge("bundlepool/bundles", nil)
+	slotsGauge  = metrics.NewRegisteredGauge("bundlepool/slots", nil)
+)
+
+// BlockChain defines the minimal set of methods needed to back a tx pool with
+// a chain. Exists to allow mocking the live chain out of tests.
+type BlockChain interface {
+	// Config retrieves the chain's fork configuration.
+	Config() *params.ChainConfig
+
+	// CurrentBlock returns the current head of the chain.
+	CurrentBlock() *types.Header
+
+	// GetBlock retrieves a specific block, used during pool resets.
+	GetBlock(hash common.Hash, number uint64) *types.Block
+
+	// StateAt returns a state database for a given root hash (generally the head).
+	StateAt(root common.Hash) (*state.StateDB, error)
+}
+
+type BundleSimulator interface {
+	SimulateBundle(bundle *types.Bundle) (*big.Int, error)
+}
+
+type BundlePool struct {
+	config Config
+	gasTip atomic.Pointer[big.Int] // Currently accepted minimum gas tip
+
+	bundles    map[common.Hash]*types.Bundle
+	bundleHeap BundleHeap
+	mu         sync.RWMutex
+
+	slots uint64 // Number of slots currently allocated
+
+	bundleGasPricer *BundleGasPricer
+	simulator       BundleSimulator
+}
+
+func New(config Config, chain BlockChain) *BundlePool {
+	// Sanitize the input to ensure no vulnerable gas prices are set
+	config = (&config).sanitize()
+
+	pool := &BundlePool{
+		config:          config,
+		bundles:         make(map[common.Hash]*types.Bundle),
+		bundleHeap:      make(BundleHeap, 0),
+		bundleGasPricer: NewBundleGasPricer(config.BundleGasPricerExpireTime),
+	}
+
+	return pool
+}
+
+func (p *BundlePool) SetBundleSimulator(simulator BundleSimulator) {
+	p.simulator = simulator
+}
+
+func (p *BundlePool) Init(gasTip *big.Int, head *types.Header, reserve txpool.AddressReserver) error {
+	// Set the basic pool parameters
+	p.reset(head)
+
+	// Since the user might have modified their pool's capacity, evict anything
+	// above the current allowance
+	for p.slots > p.config.GlobalSlots {
+		p.drop()
+	}
+
+	bundleGauge.Update(int64(len(p.bundles)))
+	slotsGauge.Update(int64(p.slots))
+
+	return nil
+}
+
+func (p *BundlePool) FilterBundle(bundle *types.Bundle) bool {
+	for _, tx := range bundle.Txs {
+		if !p.filter(tx) {
+			return false
+		}
+	}
+	return true
+}
+
+// AddBundle adds a mev bundle to the pool
+func (p *BundlePool) AddBundle(bundle *types.Bundle) error {
+	if p.simulator == nil {
+		return txpool.ErrSimulatorMissing
+	}
+
+	price, err := p.simulator.SimulateBundle(bundle)
+	if err != nil {
+		return err
+	}
+	minimalGasPrice := p.bundleGasPricer.MinimalBundleGasPrice()
+	if price.Cmp(minimalGasPrice) < 0 {
+		return txpool.ErrBundleGasPriceLow
+	}
+	bundle.Price = price
+	p.bundleGasPricer.Push(price)
+
+	hash := bundle.Hash()
+	if _, ok := p.bundles[hash]; ok {
+		return txpool.ErrBundleAlreadyExist
+	}
+	for p.slots+numSlots(bundle) > p.config.GlobalSlots {
+		p.drop()
```

Comment:
should not drop directly.. should check the gas price is higher than current one, and also no need to delete all the bundles at that price.. `p.deleteBundle(leastPriceBundleHash)` this impact too much.